### PR TITLE
odig: Init at 0.0.9

### PR DESCRIPTION
**Pull request by Julow**
###### Description of changes

This PR adds `odig`, a tool for locally building documentation of OCaml libraries and two missing dependencies:
- `ocamlPackages.buildTopkgPackage`, a function for building `topkg` packages. Very similar to `buildDunePackage`.
  This function helps building an OCaml package that builds with topkg. There are currently many such packages in nixpkgs and this function would greatly simplify adding more.
  This is heavily inspired by `ocamlPackages.buildDunePackage`.
  I did not modify the existing packages to use this new function.
- `ocamlPackages.b0`, a dependency of `odig`.

`odig` is added to `all-packages` as it is a binary tool.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/ocaml-modules/odig/default.nix
Comment (Julow) [inline pkgs/development/ocaml-modules/odig/default.nix:22]: What is the right license ? The description says:

> odig is distributed under the ISC license. The theme fonts have their own licenses.

The extra licenses in Opam are: `"LicenseRef-ParaType-Free-Font-License" "LicenseRef-DejaVu-fonts"`

**Review by Julow — COMMENTED** [review]
Ping @vbgl as you maintain other topkg packages.

**Comment by vbgl** [comment]
I find the name of `buildTopkgPackage` poorly chosen. Contrary to what you wrote, this function does not build a package: it makes a derivation. In `nix` context, “build” usually refers to what happens when `nix build` is run.

I doubt about the validity of that assertion:
>  this function would greatly simplify adding more

I’d be pleased to be proven wrong.

**Comment by Julow** [comment]
It's named after `buildDunePackage` and `buildOasisPackage`. Dune, oasis and topkg all being build systems. It's also the case in other parts of nixpkgs: `buildPythonPackage`, `buildNodePackage`.
This one can't be argued.

The "build" term comes from that it override the "builder" of a derivation ?

To prove that it's useful, first there's the popularity of `buildDunePackage`. Then, this is the description of the `fpath` library:

```nix
{ stdenv, lib, fetchurl, ocaml, findlib, ocamlbuild, topkg, astring }:

if lib.versionOlder ocaml.version "4.03"
then throw "fpath is not available for OCaml ${ocaml.version}"
else

stdenv.mkDerivation rec {
  pname = "ocaml${ocaml.version}-fpath";
  version = "0.7.3";

  src = fetchurl {
    url = "https://erratique.ch/software/fpath/releases/fpath-${version}.tbz";
    sha256 = "03z7mj0sqdz465rc4drj1gr88l9q3nfs374yssvdjdyhjbqqzc0j";
  };

  nativeBuildInputs = [ ocaml findlib ocamlbuild topkg ];
  buildInputs = [ topkg ];

  propagatedBuildInputs = [ astring ];

  strictDeps = true;

  inherit (topkg) buildPhase installPhase;

  meta = {
    description = "An OCaml module for handling file system paths with POSIX and Windows conventions";
    homepage = "https://erratique.ch/software/fpath";
    license = lib.licenses.isc;
    maintainers = [ lib.maintainers.vbgl ];
    inherit (ocaml.meta) platforms;
  };
}
```

This is with the new function:

```nix
{ lib, buildTopkgPackage, fetchurl, astring }:

buildTopkgPackage rec {
  pname = "fpath";
  version = "0.7.3";

  minimalOCamlVersion = "4.03";

  src = fetchurl {
    url = "https://erratique.ch/software/fpath/releases/fpath-${version}.tbz";
    sha256 = "03z7mj0sqdz465rc4drj1gr88l9q3nfs374yssvdjdyhjbqqzc0j";
  };

  propagatedBuildInputs = [ astring ];

  meta = {
    description = "An OCaml module for handling file system paths with POSIX and Windows conventions";
    homepage = "https://erratique.ch/software/fpath";
    license = lib.licenses.isc;
    maintainers = [ lib.maintainers.vbgl ];
  };
}
```

Many things are automated away: derivation name, topkg related inputs, build/install commands, `meta.platforms`, small things like `strictDeps`.

**Comment by ulrikstrid** [comment]
I agree with @Julow about the name, it makes sense in the context of other functions that has the same type of functionality.

It's slightly cleaner but not a huge win, but it is however nice to have the same API for different build systems across OCaml packages.

**Comment by Julow** [comment]
> It's slightly cleaner but not a huge win

Sorry to insist but I believe this is a huge win. The popularity of the other `build*Package` shows that such an abstraction is necessary.

It's reducing the length of a package definition by 1/3 and abstract away 6 concepts that are unrelated to the package and will evolve independently (topkg's deps, topkg's build rules, package naming, minimum ocaml version, `strictDeps`, `meta.platforms`).

**Comment by ulrikstrid** [comment]
Actually, I missed the pname stuff, that's much better actually. And I'm not arguing against adding this, just that it at first glance didn't look huge. But I think we should add it and then change other packages in a follow-up PR.

**Comment by Julow** [comment]
How can this debate be unstuck ?

**Review by Alizter — APPROVED** [review]

**Comment by Julow** [comment]
@vbgl It seems that this PR is blocked by you. What do you think of [my answer](https://github.com/NixOS/nixpkgs/pull/241493#issuecomment-1621218406) ?

**Comment by vbgl** [comment]
I found your answer totally unconvincing. You justify the choice of a poor name by the fact that there exist other bad names. You pretend that the `topkg` wrapper will be useful because dune is a popular build tool.

I express my personal opinion about the changes made in this PR but by no means I’ve blocked anything.

**Comment by Alizter** [comment]
If the `build*Package` name is what is bothering you, should we think about renaming them to `mk*Package`? If we decide to do so, then perhaps `mkTopkgPackage` would be a better name?

Regarding the usefulness of this new utility, what are some other derivations that could benefit from this? If we don't have any other derivations right now, then perhaps we can hold off introducing it until there is a need.

**Comment by Julow** [comment]
Hi, I just rebased. The added `buildTopkgPackage` function is critical to maintain the added package and has an approval.
Can we merge this ?

**Review by ulrikstrid — APPROVED** [review]
I think this is good to go, the topkg helper makes sense and follows convention from other such helpers

**Pull request merged**
